model/search_pipeline/search_pipeline.smithy

```diff
@@ -0,0 +1,110 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+//  The OpenSearch Contributors require contributions made to
+//  this file be licensed under the Apache-2.0 license or a
+//  compatible open source license.
+
+$version: "2"
+namespace OpenSearch
+
+map SearchPipelineMap{
+    key: String,
+    value: SearchPipelineStructure
+}
+
+structure SearchPipelineStructure{
+    version: Integer,
+    request_processors: RequestProcessorsList,
+    response_processors: ResponseProcessorsList,
+    phase_results_processors: PhaseResultsProcessorsList
+}
+
+list RequestProcessorsList {
+    member: RequestProcessors
+}
+
+list ResponseProcessorsList {
+    member: RequestProcessors
+}
+
+list PhaseResultsProcessorsList {
+    member: PhaseResultsProcessors
+}
+
+structure RequestProcessors {
+    filter_query: FilterQuery,
+    neural_query_enricher: NeuralQueryEnricher,
+    script: SearchScript
+}
+
+structure ResponseProcessors {
+    personalize_search_ranking: PersonalizeSearchRanking,
+    rename_field: RenameField,
+}
+
+structure PhaseResultsProcessors {
+    normalization: ScoreNormalization,
+    combination: ScoreCombination,
+    tag: String,
+    description: String,
+    ignore_failure: Boolean
+}
+
+structure ScoreNormalization{
+    technique: String
+}
+
+structure ScoreCombination{
+    technique: String,
+    parameters: ScoreWeights
+}
+
+list ScoreWeights{
+    member: Float
+}
+
+structure FilterQuery {
+    query: UserDefinedObjectStructure,
+    tag: String,
+    description: String,
+    ignore_failure: Boolean
+}
+
+structure NeuralQueryEnricher {
+    default_model_id: String,
+    neural_field_default_id: NeuralFieldMap
+    tag: String,
+    description: String,
+}
+
+structure SearchScript {
+    source: String,
+    lang: String,
+    tag: String,
+    description: String,
+    ignore_failure: Boolean
+}
+
+structure PersonalizeSearchRanking {
+    campaign_arn: String,
+    recipe: String,
+    weight: Float,
+    item_id_field: String,
+    iam_role_arn: String,
+    tag: String,
+    description: String,
+    ignore_failure: Boolean
+}
+
+structure RenameField {
+    field: String,
+    target_field: String,
+    tag: String,
+    description: String,
+    ignore_failure: Boolean
+}
+
+map NeuralFieldMap {
+  key: String,
+  value: String
+}
```

Comment:
I've just gone through and reorganized the ordering of the structures and fixed some things:
- Added missing processors added in `2.12.0`
- Removed pluralization from list and structure naming of the processors
- Swapped relevant structures to unions
- Corrected some naming & added relevant `@required` traits

```suggestion
map SearchPipelineMap {
    key: String
    value: SearchPipelineStructure
}

structure SearchPipelineStructure {
    version: Integer
    request_processors: RequestProcessorList
    response_processors: ResponseProcessorList
    phase_results_processors: PhaseResultsProcessorList
}

// === Common Properties ===

@mixin
structure ProcessorBase {
    tag: String
    description: String
}

@mixin
structure IgnoreableProcessorBase with [ProcessorBase] {
    ignore_failure: Boolean
}

// === Request Processors ===

list RequestProcessorList {
    member: RequestProcessor
}

union RequestProcessor {
    filter_query: FilterQueryRequestProcessor
    neural_query_enricher: NeuralQueryEnricherRequestProcessor
    script: SearchScriptRequestProcessor
    oversample: OversampleRequestProcessor
}

structure FilterQueryRequestProcessor with [IgnoreableProcessorBase] {
    query: UserDefinedObjectStructure
}

structure NeuralQueryEnricherRequestProcessor with [ProcessorBase] {
    default_model_id: String
    neural_field_default_id: NeuralFieldMap
}

map NeuralFieldMap {
    key: String
    value: String
}

structure SearchScriptRequestProcessor with [IgnoreableProcessorBase] {
    @required
    source: String

    lang: String
}

structure OversampleRequestProcessor with [IgnoreableProcessorBase] {
    @required
    sample_factor: Float

    content_prefix: String
}

// === Response Processors ===

list ResponseProcessorList {
    member: RequestProcessor
}

union ResponseProcessor {
    personalize_search_ranking: PersonalizeSearchRankingResponseProcessor
    retrieval_augmented_generation: RetrievalAugmentedGenerationResponseProcessor
    rename_field: RenameFieldResponseProcessor
    rerank: RerankResponseProcessor
    collapse: CollapseResponseProcessor
    truncate_hits: TruncateHitsResponseProcessor
}

structure PersonalizeSearchRankingResponseProcessor with [IgnoreableProcessorBase] {
    @required
    campaign_arn: String

    @required
    recipe: String

    @required
    weight: Float

    item_id_field: String

    iam_role_arn: String
}

structure RetrievalAugmentedGenerationResponseProcessor with [ProcessorBase] {
    @required
    model_id: String

    @required
    context_field_list: ContextFieldList

    system_prompt: String

    user_instructions: String
}

list ContextFieldList {
    member: String
}

structure RenameFieldResponseProcessor with [IgnoreableProcessorBase] {
    @required
    field: String

    @required
    target_field: String
}

structure RerankResponseProcessor with [IgnoreableProcessorBase] {
    ml_opensearch: MLOpenSearchReranker
    context: RerankContext
}

structure MLOpenSearchReranker {
    @required
    model_id: String
}

structure RerankContext {
    @required
    document_fields: DocumentFieldList
}

list DocumentFieldList {
    member: String
}

structure CollapseResponseProcessor with [IgnoreableProcessorBase] {
    @required
    field: String

    context_prefix: String
}

structure TruncateHitsResponseProcessor with [IgnoreableProcessorBase] {
    target_size: Integer
    context_prefix: String
}

// === Phase Results Processors ===

list PhaseResultsProcessorList {
    member: PhaseResultsProcessor
}

union PhaseResultsProcessor {
    @jsonName("normalization-processor")
    normalization_processor: NormalizationPhaseResultsProcessor
}

structure NormalizationPhaseResultsProcessor with [IgnoreableProcessorBase] {
    normalization: ScoreNormalization
    combination: ScoreCombination
}

structure ScoreNormalization {
    technique: ScoreNormalizationTechnique
}

enum ScoreNormalizationTechnique {
    MIN_MAX = "min_max"
    L2 = "l2"
}

structure ScoreCombination {
    technique: ScoreCombinationTechnique
    parameters: ScoreWeights
}

enum ScoreCombinationTechnique {
    ARITHMETIC_MEAN = "arithmetic_mean"
    GEOMETRIC_MEAN = "geometric_mean"
    HARMONIC_MEAN = "harmonic_mean"
}

structure ScoreCombinationParameters {
    weights: ScoreWeights
}

list ScoreWeights {
    member: Float
}

```